src/python/pants/core/util_rules/external_tool.py

```diff
@@ -0,0 +1,248 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import List
+
+from pants.base.build_environment import get_buildroot
+from pants.core.util_rules.archive import ExtractedDigest, MaybeExtractable
+from pants.engine.fs import Digest, DirectoryToMaterialize, Snapshot, UrlToFetch
+from pants.engine.platform import Platform, PlatformConstraint
+from pants.engine.rules import RootRule, rule
+from pants.engine.selectors import Get
+from pants.subsystem.subsystem import Subsystem
+from pants.util.memo import memoized_method
+from pants.util.meta import classproperty
+
+
+class UnknownVersion(Exception):
+    pass
+
+
+class ExternalToolError(Exception):
+    pass
+
+
+@dataclass(frozen=True)
+class ExternalToolRequest:
+    url_to_fetch: UrlToFetch
+    exe: str
+
+
+@dataclass(frozen=True)
+class DownloadedExternalTool:
+    digest: Digest
+    exe: str
+
+
+class ExternalTool(Subsystem):
+    """Configuration for an invocable tool that we download from an external source.
+
+    Subclass this to configure a specific tool.
+
+
+    Idiomatic use:
+
+    class MyExternalTool(ExternalTool):
+        options_scope = "my-external-tool"
+        default_version = "1.2.3"
+        default_known_versions = [
+          "1.2.3|darwin|deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef|222222",
+          "1.2.3|linux |1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234abcd|333333",
+        ]
+
+        @classmethod
+        def generate_url(cls, plat: Platform, version: str) -> str:
+            ...
+
+        @classmethod
+        def generate_exe(cls, plat: Platform, version: str) -> str:
+            ...
+
+    @rule
+    def my_rule(my_external_tool: MyExternalTool):
+      downloaded_tool = await Get[DownloadedExternalTool](
+        ExternalToolRequest, my_external_tool.get_request(Platform.current)
+      )
```

Comment:
It would be pretty cool to have this in the official engine README.md. I should also get #9446 in.